### PR TITLE
feat(gui): Runtime Monitor Memory tab — process memory over time (#2554)

### PR DESCRIPTION
**Pull request by skerker**
## Summary

Part of #2554 — the Memory tab, the third increment of the Runtime Monitor after Threads and Logs (#5246). The dialog gains a **Memory** tab: a summary line naming what "resident" means on this platform and how many samples the history holds, four readouts (Resident, Peak, Private, Virtual), a timeframe selector (1 min / 5 min / 15 min / 1 hour) and a chart of Resident, Private and Peak over the selected window on the same `TimeSeriesGraphWidget` the network dialog uses.

The reading comes from the collector the Threads tab already runs on its own thread at 1.5 s: each tick now also emits a `MemorySample` (a flat copy of `ProcessMemorySnapshot::capture()`, #4293's telemetry) queued to the GUI, before the thread-table enumeration and its early return, so a platform whose thread enumeration fails still gets a Memory tab. The history is a bounded gui-side ring (`MemoryHistoryRing`, 2400 samples = one hour raw; 1 s points to five minutes, bucket averages of max(5 s, range/300) beyond, placed at the bucket centre as `NetworkDiagnosticsDialog::updateCharts()` places its own). `MainWindow` owns the ring and hands it to the dialog, because the dialog is `WA_DeleteOnClose` and a trend chart that forgot everything on Close would not be a trend: Close and reopen shows what was sampled before, with a break in each line where nothing was sampled.

What the platform did not measure is shown as a dash, never as `0.0 MB`: Windows never reports virtual bytes, Linux reports no private bytes without `/proc/self/smaps_rollup`, and an invalid sample reports nothing — those readouts read "—", the chart skips them, and an invalid sample's summary names no metric. Acceptance criterion 4 ("the chart must move down as well as up") holds: resident memory is the current footprint, not a peak, and the demo run below shows it dropping.

**Deliberate choices, disclosed:**
- The collector samples **only while the dialog is open** (an always-on enumeration every 1.5 s is observer effect on the thing being measured — the question left open in [my 08-25 comment](https://github.com/aethersdr/AetherSDR/issues/2554#issuecomment-5416915052); a maintainer who wants always-on history can start the collector from `MainWindow` instead, the ring is already there). Samples accrue only while open; the "N samples" count is the ring's size.
- The timeframe selector lives on the Memory tab, not at the window level as the issue sketches: Threads has a fixed 60 s window and Logs none, so a window-level control would be blank on two of three tabs. It is not persisted, like the network dialog's.
- Not in this increment: the major page-fault rate (no counter in `MemoryTelemetry`), the build-flagged heap breakdown, and the seven-day compacting history — those stay with the Overview increment, as in my 08-27 comment.
- On macOS, Peak is `resident_size_peak` while Resident is `phys_footprint` (#4293's accounting); the two are not the same metric, which is why Peak reads well above Resident there — the readout's tooltip says so.
- Beyond five minutes the newest partial bucket is not drawn until its 5 s / 12 s edge passes; that is the network dialog's identical rule, kept so the two charts read alike. The readouts always show the newest sample.
- One shared-widget change: `TimeSeriesGraphWidget` now computes its left-gutter value hints before the axis ticks and does not paint a tick label that would sit under a hint (previously the label was painted and then partially covered — overlapping digits, visible with three series close together). `NetworkDiagnosticsDialog` inherits this: a tick label within 18 px of a hint is no longer painted there either.
- No new gui→core header include (AGENTS.md): the ring keeps its own record and its own copy of the 1.5 s cadence, pinned to the collector's by a `static_assert` in `SystemInfoDialog.cpp`; the collector include moves from `SystemInfoDialog.cpp` to `.h`. `tools/gen_touchpoint_manifest.py --check` is up to date on this branch.

`CHANGELOG.md` is not touched (release-prep file).

## Constitution principle honored

Principle VIII — Evidence over assertion: every behavioural claim above is backed by a measured run on the demo simulator or a test that fails when the behaviour is removed; the "dash, not 0.0 MB" rule exists because the tab must not present a number the platform never produced.

## Test

Three socket-free targets, declared in `tests/tests.cmake` (none runs in a PR CI lane; local results below):

- `memory_history_ring_test` — retention and eviction order (which sample survives), raw window slicing at both edges, the bucket rule shared with the network dialog, exact bucket averages on full buckets and bucket-centre placement, the line-break threshold per timeframe, and that invalid or zero-valued records are not plotted. All samples constructed (arithmetic and routing only).
- `system_info_collector_test` — the real thread wiring (parentless collector moved to a `QThread`, `init()` on `started`, the 1.5 s timer): a memory sample arrives on the caller's thread within 5 s, is timestamped between start and receipt, names one of the snapshot's platform metrics, and on a supported platform is valid with non-zero resident bytes. Measured on the live test process; the peak ≥ resident check is skipped on macOS where the two are different metrics.
- `system_info_dialog_test` (existing target, extended) — the Memory tab's widgets and defaults, readouts and summary driven through `applyMemorySample()` with constructed samples, the chart moving down, timeframe changes, hide/show keeping the history, a Windows-shaped sample (shape from `MemoryTelemetry.cpp`'s Windows branch, values constructed) reading a dash for the unset field, an invalid sample reading four dashes and a metric-less summary, the accessible-name announcement, and two dialogs sharing one injected ring across the first's destruction.

Mutation checks (each restored with the inverse edit, `git diff --stat` empty, rebuilt green): collector emit dropped → arrival check fails; ring eviction disabled → 3 failures; bucket averaging replaced by last-sample-wins → 1 failure; bucket centre moved to the start → 1; unplottable rule forced true → 2; dialog ring push skipped → 6; dash rule removed → 3; injected ring ignored → 2.

Full local `ctest` on `b15b6a60`, the whole registered suite of 341: 335 passed, 3 skipped, 3 failed — `bridge_docs_check` (stale verb table) and `phone_tx_filter_numeric_entry_test` (subprocess abort under `QT_QPA_PLATFORM=offscreen`) fail identically on the merge base; `hl2_state_restore_test` is #5394, reported failing on `main` at `10a847b7` the same day. None touches this change.

## Proof — agent automation bridge, demo simulator

`b15b6a60`, clean RelWithDebInfo build (macOS Intel, Qt 6.8.3; About-dialog SHA checked), `DEMO-0001`, isolated settings store, TX automation off; readouts are the tab's labels read back over the bridge (`dumpTree`), the count is the summary line.

| Step | Reading |
|---|---|
| Open Help → Runtime Monitor → Memory | 27 samples at the first reading (~40 s open), three lines at the right edge, Timeframe 5 minutes |
| Close, wait 12 s, reopen | 35 → 46 → 57: the count continues; the chart shows the earlier segment, a gap, the new segment |
| Disconnect with the tab open | 67 → 89 samples, radio `connected=false`: sampling does not depend on the radio |
| Reconnect | Resident steps down (314.4 → 278.0 MB) — the chart moves down as well as up |
| Timeframe 1 minute / 15 minutes / 1 hour | three lines drawn in each; 1 minute shows the reconnect drop, 15 minutes and 1 hour the two segments and the gap; readouts unchanged by the switch (278.2 / 278.7 / 278.9 MB, the newest sample each time) |
| Quit with the dialog open | process exits; no crash report in `~/Library/Logs/DiagnosticReports` |

Earlier heads, same rig, one line each: `f089456a` (pre-review) — close/reopen 37 → 51 → 71 samples; disconnect 71 → 87 → 102; footprint 323 → 287 MB on reconnect; 1-minute, 5-minute and 1-hour views with 784 samples. `7d886163` — the tick-label fix confirmed on the 1-hour view where the digits had overlapped.

Screenshots below: the About dialog with the SHA, the Memory tab after close/reopen (the earlier segment, the gap, the new segment), and the 1-hour view. Evidence bundle `aethersdr-pr5397-memory-tab-2026-09-03.zip` (attached below; sha256 `2be257bd34d03ee9234210a6aae2e13c121f45b21b623918f8dd7b1287ecc6f1`): the run transcript with every readout, the whole unedited ctest log, the app log and the five captures, with a README mapping each claim above to its file. Sanitized; no radio, no RF.

aethersdr-2554-about-b15b6a60-2026-09-03

aethersdr-2554-memory-tab-after-reopen-b15b6a60-2026-09-03

aethersdr-2554-memory-tab-1hour-b15b6a60-2026-09-03

[aethersdr-pr5397-memory-tab-2026-09-03.zip](https://github.com/user-attachments/files/31795390/aethersdr-pr5397-memory-tab-2026-09-03.zip)

## What I tried to break

- Teardown order: `MainWindow` destroys the ring before its child dialog; the dialog's destructor and hide path never touch the ring (read), and quitting with the dialog open left no crash report (run).
- The queued sample after hide is dropped by the sampling-generation guard; a dialog given no ring uses its own; two dialogs sharing one ring across the first's destruction (test).
- Window edges inclusive at 1 s resolution (41 points for 60 s at 1.5 s), no negative x for any offered range, bucket-aligned cutoffs; wall-clock steps and sleep/wake reasoned only — the x axis is wall time on purpose, so a 3 h sleep shows as a gap, not spliced shut.
- Every `MemoryTelemetry.cpp` platform branch read for which fields it leaves at zero and when `valid` is false — that reading is where the dash rule came from.
- Not driven: Windows (`workingSet`) and Linux (`VmRSS`) readouts on their own platforms — reasoned from the telemetry source and pinned by the shaped samples in the dialog test.

## Test plan

- [x] Local build passes (`cmake --build build`) — clean RelWithDebInfo, macOS Intel, Qt 6.8.3, every head
- [x] Behavior verified on a real radio if applicable — **n/a:** process memory of the client itself; the demo run shows sampling is independent of the radio connection
- [x] Existing tests pass (CI); the three targets above run locally: pass, and fail under the listed mutations
- [x] Reproduction steps documented if user-reported bug — **n/a, feature increment**

## Checklist

- [x] Commits are signed (`docs/COMMIT-SIGNING.md`) — SSH
- [x] No new flat-key `AppSettings` calls — **n/a, no settings touched** (the timeframe is not persisted, like the network dialog's)
- [x] Code is clean-room — yes
- [x] All meter UI uses `MeterSmoother` — **n/a, no meter UI** (readouts are 1.5 s samples, not meters)
- [x] Documentation updated if user-visible behavior changed — `docs/architecture/pipelines.md`: the collector-thread line now says it samples CPU and process memory. No user doc covers the Runtime Monitor yet. `CHANGELOG.md` not touched per the template
- [x] Security-sensitive changes reference a GHSA — **n/a**

— authored by agent (Claude Code) on behalf of @skerker

### CODE REVIEW

**Comment by skerker** [comment]
**Linux leg of the Memory tab smoke** — the reading the PR body lists as not driven on its own platform (`VmRSS`).
Clean build of `b15b6a60` on Linux (Ubuntu 24.04.4, kernel 7.0, GCC 13.3, Qt 6.8.3, X11), Build SHA captured at
configure time and confirmed in Help → About before any reading. Demo radio only, isolated settings store, TX
automation off; readouts are the tab's labels read back over the bridge (`dumpTree`). Verdict: **all six cues pass;
Linux fills all four readouts; one rendering observation to disclose.**

**The platform reading.** `/proc/self/smaps_rollup` exists on this box, so Private is populated: summary line
`Resident = VmRSS · 33 samples`, readouts `489.1 MB` / `502.8 MB` / `411.0 MB` / `15800.8 MB`
(Resident / Peak / Private / Virtual). No dash and no `0.0 MB` anywhere.

**The tab.**
- Close ~10 s, reopen: the count carried on (34 → 45 → 54) with nothing sampled during the closed interval, and the
  chart shows segment · gap · segment.
- Disconnect with the tab open: sampling continued (70 → 83 samples at `radio.connected = false`).
- Reconnect: Resident stepped **up** here, 488.1 → 506.9 MB (the macOS run saw a step down); Peak rose 502.8 → 535.6 MB.
- 1 minute / 15 minutes / 1 hour: all three series drawn in each view, readouts unchanged by the switch.
- Quit with the dialog open on the Memory tab: exit code 0, no crash artifact.

**Tests.** The three related targets (`memory_history_ring_test`, `system_info_collector_test`,
`system_info_dialog_test`) pass 3/3. Full registered `ctest` at the PR head: **333/339 passed, 3 skipped, 3 failed** —
`bridge_docs_check` (stale verb table), `hl2_state_restore_test` (#5394), `phone_tx_filter_numeric_entry_test`
(Qt testlib assert) — the same trio the macOS run reported; none touches this PR's files or tests.

**One observation (cosmetic, consistent with the review's nit 2):** on every view the `500 MB` axis tick label is
suppressed under the `507`/`536` value hints, so that band of the left axis shows only the hints. The graph stays
readable; noting it because it is the shared-widget change seen rendered rather than reasoned from geometry.

Screenshots below: the Memory tab on first open, and the 1-hour view. Evidence bundle
`aethersdr-pr5397-linux-memory-tab-2026-09-03.zip` (attached below; sha256
`eb3caa74662af4b1467ee862f5930477a7b1f50976e2a564a4b8d4af04369fe5`): the run transcript with every readout and
timestamp, the whole unedited ctest log, the configure log, box facts, and six captures, with a README mapping each
claim above to its file. Sanitized; no radio hardware, no RF.

aethersdr-pr5397-linux-memory-tab-cue1-2026-09-03

aethersdr-pr5397-linux-memory-tab-cue5-1hour-2026-09-03

[aethersdr-pr5397-linux-memory-tab-2026-09-03.zip](https://github.com/user-attachments/files/31801645/aethersdr-pr5397-linux-memory-tab-2026-09-03.zip)




— authored by agent (Claude Code) on behalf of @skerker

**Comment by skerker** [comment]
**Windows leg of the Memory tab smoke** — the reading the PR body lists as not driven on its own platform (Virtual
must read a dash, never `0.0 MB`, on a platform that leaves the field unset).
Clean build of `b15b6a60` on Windows (Windows 11 Pro build 26200 x64, MSVC 19.44, CMake 3.31.6, Ninja 1.13.2,
Qt 6.8.3 msvc2022_64, RelWithDebInfo), Build SHA captured at configure time and confirmed in Help → About before any
reading; 0 MSVC warnings in any file the PR touches. Demo radio only, isolated settings store, TX automation off;
readouts are the tab's labels read back over the bridge (`dumpTree`). Verdict: **all six cues pass; Windows shows the
dash for Virtual and "working set" as the resident metric; one observation to disclose.**

**The platform reading.** Summary line `Resident = working set · 37 samples`, readouts `421.2 MB` / `532.3 MB` /
`546.5 MB` / `—` (Resident / Peak / Private / Virtual). The dash is a dash, not `0.0 MB`. Private sits above Resident
here (private bytes counts committed pages that are not in the working set), which the tooltips' wording already
allows for.

**The tab.**
- Close ~12 s, reopen: the count carried on (37 → 57 → 65) with nothing sampled during the closed interval, and the
  chart shows segment · gap · segment.
- Disconnect with the tab open: sampling continued (65 → 89 samples at `radio.connected = false`).
- Reconnect: Resident stepped **up** here, 408.9 → 420.0 MB (Linux saw up, macOS saw down); Peak unchanged at 532.3 MB.
- 1 minute / 15 minutes / 1 hour: all three series drawn in each view, readouts unchanged by the switch.
- Quit with the dialog open on the Memory tab: orderly shutdown (`aether.shutdown: phase=application event=complete`
  is the log's last line), no Application-log error and no crash dump for the run.

**Tests.** The three related targets (`memory_history_ring_test`, `system_info_collector_test`,
`system_info_dialog_test`) pass 3/3 (they need the Qt bin directory on `PATH` to load). Full registered `ctest` at the
PR head: **327/339 passed, 3 skipped, 9 failed.** Six are the Windows-main baseline set this box has reported since 08-27
(`crdv_manifest_test`, `dss_renderer_test`, `aether_mcp_field_mapping`, `hl2_state_restore_test` = #5000,
`connection_panel_size_test`, `phone_tx_filter_numeric_entry_test`), plus `vkamp_connection_test` (characterised 08-27)
and `bridge_docs_check` (the stale verb table the macOS and Linux runs also hit). One is new to this box:
`ctcss_tone_label_test` (added 09-01 by #5259) fails deterministically here, 3/3 in isolation, on
"an uncoded row paints frequency ink and leaves the designation column blank" under the offscreen platform — it did not
exist at the last Windows baseline, and nothing this PR touches is involved. None of the nine touches this PR's files or
tests; the three related targets also pass inside the full run.

**One observation (cosmetic, consistent with the review's nit 2 and the Linux report):** on every view the `500 MB`
axis tick label is suppressed under the `543`/`532` value hints. The graph stays readable.

Screenshots below: the Memory tab on first open, and the 1-hour view. Evidence bundle
`aethersdr-pr5397-windows-memory-tab-2026-09-03.zip` (attached below; sha256 `420e3fdfb7a5e827b10c11f10e19692ae631999a4ec92602375f4102d8d34682`): the run transcript with every
readout and timestamp, the whole unedited ctest log, the configure and build logs, the app log, the Event Viewer check,
box facts, and six captures, with a README mapping each claim above to its file. Sanitized; no radio hardware, no RF.


[aethersdr-pr5397-windows-memory-tab-2026-09-03.zip](https://github.com/user-attachments/files/31805316/aethersdr-pr5397-windows-memory-tab-2026-09-03.zip)

aethersdr-pr5397-memory-tab-1-windows-2026-09-03

aethersdr-pr5397-memory-tab-5c-1hour-windows-2026-09-03

— authored by agent (Claude Code) on behalf of @skerker

**Review by Ozy311 — APPROVED** [review]
This delivers the Memory-tab portion of #2554: process-memory trends, live readouts, four selectable timeframes, and history retained across dialog close/reopen. **Approve with the existing non-blocking nits; no new blocking findings.** Reviewed exact head `b15b6a60af0ef1df9ea00786188d6f1c279702bf`.

| File group | Change and scope |
|---|---|
| SystemInfoCollector + pipelines documentation | Publishes memory samples before the thread-enumeration early return; in scope, no added collection thread. |
| MemoryHistoryRing | Bounded GUI-owned history and window slicing; in scope. |
| MainWindow + SystemInfoDialog | Owns retained history and supplies the Memory tab/readouts/timeframe controls; in scope. |
| TimeSeriesGraphWidget | Suppresses tick labels underneath live-value hints; supports this feature and also affects Network Diagnostics. Shared impact checked below. |
| Three focused tests + tests.cmake | Socket-free tests and registration; in scope. |

**Blockers:** None.

**Existing non-blocking notes:** The newest partial bucket can leave the long-window plot behind the live readout; wall-clock corrections remain an untested cosmetic edge case. The shared graph impact and the absence of these three tests from the filtered CI lanes remain relevant disclosures. Existing inline threads already cover the rendering/bucket concerns, so this review does not duplicate them.

**Empirical validation and attempts to break it**

- Built the exact PR head as a native macOS arm64 application with Qt 6.11.1, RelWithDebInfo. All three focused socket-free tests passed: `memory_history_ring_test`, `system_info_collector_test`, and `system_info_dialog_test` (3/3, 2.65 seconds). Existing ring tests failed as expected after separately introducing wrong averaging and wrong-end eviction into copied headers outside the repository; the PR source was unchanged.
- Signed the isolated app with the operator's established AetherSDR Local Development certificate; strict deep codesign verification passed. The native Cocoa instance used private settings, an authenticated explicit automation socket, and disabled transmit. Immediately after bridge identity verification, PC output mute was explicitly set and read back as `muted:true`; every subsequent audio snapshot remained muted.
- Drove the app through the automation bridge, using native computer use for the painted legend. Approximately 361 seconds of native validation produced 177 retained samples. Exercised all four timeframes, dark/light themes, all memory readouts and accessible values, and Resident legend isolate/restore. Observed genuine descending resident values (339.3 → 294.0 → 260.8 MB), with corresponding downward traces.
- Closed the dialog for 30.94 seconds and 61.08 seconds; reopening preserved exactly 95 and 119 samples respectively before the next tick. Sampling resumed, and the five-minute/hour views showed the dialog-close sampling gaps. Three further close/reopen cycles retained history. Normal application shutdown with Runtime Monitor open exited 0; no thread/timer ownership or assertion failure appeared in the captured log.
- Initial validation was disconnected; the only later connection was muted `DEMO-0001` for Network Diagnostics. Its latency/audio-buffer charts populated, but stream-rate counters stayed zero. Crowded-rate rendering therefore used the actual base/head shared widget with constructed populated six-series, log-scale data at 900 and 480 px. The head removed tick-label fragments underneath hints while retaining series, gridlines, and remaining axis labels. Narrow-width legend truncation existed on both sides.

**Limits:** No native Linux/Windows runtime, sanitizer run, real-radio/RF test, hour-long soak, or host-clock manipulation. Missing-platform-field cases and long-window arithmetic were tested with constructed inputs. The bridge does not expose the Memory ring's raw plotted arrays; exact arithmetic is covered by the focused tests. Optional DFNR/ONNX components were unavailable locally. Four CI checks are green, including platform builds; the three focused test results above are actual local execution, not inferred from CI.

Trusted main governance rechecked at `50d70ebe9fdc4d6f01171b5cb17479bc89537be2`: manifest PASS, unchanged from the rules used for this review. Immediately before submission the head is unchanged, Ozy311 is the sole assignee, and all ten commits have valid GitHub signatures. Required human CODEOWNERS review still applies. The owned test app has stopped; no implementation changes or merge were performed.

File: src/gui/MemoryHistoryRing.h [hunk 1]
```diff
@@ -0,0 +1,169 @@
...
+    // older than the window are skipped; beyond five minutes samples sharing a
+    // bucket are averaged so a one-hour window is ~300 points, not 2400.
+    QVector<QPointF> series(Field field, int rangeSeconds, qint64 nowMs) const
+    {
+        QVector<QPointF> points;
+        if (m_samples.empty() || rangeSeconds <= 0) {
+            return points;
+        }
+        const qint64 bucketMs = bucketMsFor(rangeSeconds);
+        const qint64 endMs = bucketMs <= 1000 ? nowMs : (nowMs / bucketMs) * bucketMs;
```
[the remark this answers is not in the document]
Comment (Ozy311) [reply src/gui/MemoryHistoryRing.h:94]: Resolving as an acknowledged non-blocking limitation, consistent with the existing approval on unchanged head `b15b6a60af0ef1df9ea00786188d6f1c279702bf`. The PR body explicitly discloses that the newest partial bucket is omitted on the 15-minute and 1-hour views while the readouts show the newest sample. The brief initial empty chart and the chart/readout lag remain; no rounding change or code fix is claimed. This closes the discussion under the approved scope. See the [existing review](https://github.com/aethersdr/AetherSDR/pull/5397#pullrequestreview-5119781030).

File: src/gui/TimeSeriesGraphWidget.h [hunk 2]
```diff
@@ -180,6 +180,57 @@ class TimeSeriesGraphWidget : public QWidget {
             maxY = niceCeiling(maxY);
         }
 
+        // Per-series "last sample" hints in the left gutter.  Each
+        // visible series gets a colored label at the y-pixel matching
+        // its most recent value; labels are spread vertically to avoid
+        // overlap when several streams sit close together (e.g. RX and
+        // Audio both around ~1 Mbps).  Positioned HERE, before the axis
```
[the remark this answers is not in the document]
Comment (Ozy311) [reply src/gui/TimeSeriesGraphWidget.h:187]: Addressed by the explicit shared-widget disclosure in the PR body: Network Diagnostics inherits tick-label suppression within 18 px of a live-value hint. The [existing review](https://github.com/aethersdr/AetherSDR/pull/5397#pullrequestreview-5119781030) records base/head rendering checks with populated six-series log-scale data at 900 and 480 px: the head removed tick-label fragments while retaining series, gridlines, and remaining axis labels. Head remains `b15b6a60af0ef1df9ea00786188d6f1c279702bf`. Resolving this scope/disclosure note; no additional code change.
